src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingExtensionMethods.cs

```diff
@@ -0,0 +1,132 @@
+using Microsoft.SqlTools.ServiceLayer.Scripting.Contracts;
+using Microsoft.SqlTools.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer.Scripting
+{
+    public static class ScriptingExtensionMethods
+    {
+        public static List<ScriptingObject> GetDatabaseObjects(this SqlScriptPublishModel publishModel)
+        {
+            string serverName = null;
+            string databaseName = null;
+            return GetDatabaseObjects(publishModel, out serverName, out databaseName);
+        }
+
+        public static List<ScriptingObject> GetDatabaseObjects(this SqlScriptPublishModel publishModel, out string serverName, out string databaseName)
+        {
+            List<ScriptingObject> databaseObjects = new List<ScriptingObject>();
+            serverName = null;
+            databaseName = null;
+            bool serverAndDatabaseInitialized = false;
+
+            IEnumerable<DatabaseObjectType> objectTypes = publishModel.GetDatabaseObjectTypes();
+            foreach (DatabaseObjectType objectType in objectTypes)
+            {
+                IEnumerable<KeyValuePair<string, string>> databaseObjectsOfType = publishModel.EnumChildrenForDatabaseObjectType(objectType);
+
+                Logger.Write(
+                    LogLevel.Normal,
+                    string.Format(
+                        "Loaded SMO urn object count {0} for type {1}, urns: {2}",
+                        objectType,
+                        databaseObjectsOfType.Count(),
+                        string.Join(", ", databaseObjectsOfType.Select(p => p.Value))));
+
+                foreach (KeyValuePair<string, string> databaseObjectOfType in databaseObjectsOfType)
+                {
+                    if (!serverAndDatabaseInitialized)
+                    {
+                        Urn urn = new Urn(databaseObjectOfType.Value);
+                        serverName = urn.GetNameForType("Server");
+                        databaseName = urn.GetNameForType("Database");
+                        serverAndDatabaseInitialized = true;
+                    }
+
+                    databaseObjects.Add(new Urn(databaseObjectOfType.Value).ToScriptingObject());
+                }
+            }
+
+            return databaseObjects;
+        }
+
+        public static bool IsOperationCanceledException(this Exception e)
+        {
+            Exception current = e;
+            while (current != null)
+            {
+                if (current is OperationCanceledException)
+                {
+                    return true;
+                }
+
+                current = current.InnerException;
+            }
+
+            return false;
+        }
+
+        public static Urn ToUrn(this ScriptingObject scriptingObject, string server, string database)
+        {
+            if (scriptingObject == null)
+            {
+                throw new ArgumentNullException("scriptingObjectc");
+            }
+
+            if (string.IsNullOrWhiteSpace(server))
```

Comment:
I updated all null checks to use the Validate class. I think the default messages should be fine.

---
In reply to: [111653103](https://github.com/Microsoft/sqltoolsservice/pull/316#discussion_r111653103) [](ancestors = 111653103)